add a single method to override to customize cache keys

bkudria opened the pull request:
This commit adds a `modified_cache_key` method (on the instance and on the class level) that by default returns it's single argument unmodified. By passing all generated cache keys through this method, we expose a single point of dependency injection for cache keys. In my codebase, it's necessary sometimes to invalidate the cache for a particular model, and overriding this method to append a cache version allows for this functionality.

(On first glance `cacheable_table_name` provides this functionality, but it's used for other purposes elsewhere and is therefore unsuitable for dependency injection.)

I've omitted tests in this PR because I'm not the most experienced with RSpec and couldn't figure out a way to test this functionality without duplicating all the specs for this module. This functionality should definitely be tested and I'd welcome suggestions on how to test in a DRY manner.

bkudria committed:
Add 'modified_cache_key' for cache key DI.

By overriding the single method 'modified_cache_key', downstream code
can inject dependencies into cache key creation. This is useful for
cache versioning, for example. Example included in the README.

This commit also switches the spec suite from mocha to rr, because rr
allows stub method implementation

ScotterC commented:
Interesting.  Thanks Ben!  I'd also like an example in the readme for your use case.  

Let me see if I can get this straight.  By modifying the `modified_cache_key` further down, you can restrict what does and does not get expired for a particular model?  [`expire_model_cache`](https://github.com/flyerhzm/simple_cacheable/blob/master/lib/cacheable/expiry.rb#L3)  does not suit your needs?

bkudria commented:
No, by overriding modified_cache_key, I can change what cache_keys are. So, for example, my version looks like this:

``` ruby
def modified_cache_key(key)
  [key, self.cache_version] * '/'
end
```

Where `cache_version` is a number associated with a model. If I need to re-cache for all instances of a particular model, I just increment its cache_version. As far as I understand, expire_model_cache doesn't expire individual instance caches, right? Furthermore, I use cache_version for caching other things, and it's a useful mechanism for clearing all cached things for a model.

That's just my use-case, though. I could imagine other users using it for namespacing, or including other things in the cache key, for example. In general it seems to me this important bit of functionality should be exposed. (An alternative solution might be to depend on `cache_key` as provided by ActiveRecord. That'd be my preferred method.)

ScotterC commented:
`expire_model_cache` should completely expire caches for an instance and it's associations, methods etc.  

Still, a single avenue to modify the cache keys makes a lot of sense, but we will need to figure out a good way to test it as you noted above.  

I'd add a block of specs to each of the [cache types](https://github.com/flyerhzm/simple_cacheable/tree/master/spec/cacheable/types) which would redo one of the simpler specs with a different cache key and check `Rails.cache` directly to make sure the existence is correct and that it doesn't exist in the original key

bkudria commented:
Right - expire_model_cache is instance-level. My change enables expiring the cache for _all_ instances, past and future, all at once. If I deploy a code change to a model, and I want to invalidate the cache, looping through every model and calling `expire_model_cache` isn't feasible.

I'll write some tests and add an example in the README.

ScotterC commented:
Oh!  That makes a whole lot of sense. I've gotten used to treating the
cache as completely volatile which has blinded me to partial sweeping
expirations. Thanks.

On Thursday, February 27, 2014, Benjamin Kudria notifications@github.com
wrote:

> Right - expire_model_cache is instance-level. My change enables expiring
> the cache for _all_ instances, past and future, all at once. If I deploy
> a code change to a model, and I want to invalidate the cache, looping
> through every model and calling expire_model_cache isn't feasible.
> 
> I'll write some tests and add an example in the README.
> 
> ## 
> 
> Reply to this email directly or view it on GitHubhttps://github.com/flyerhzm/simple_cacheable/pull/52#issuecomment-36306766
> .

## 

## 

Scott Carleton
Artsicle, Co-Founder / CTO
scott@artsicle.com
C. 914 400 5085

bkudria commented:
Ok, I've updated the README with an example, and added a spec for each cache type. I also had to swap out mocha for rr, because mocha doesn't allow supplying a custom implementation for a method stub.

bkudria commented:
Not sure why Travis is complaining, the Gemfile.lock is up-to-date.

ScotterC commented:
Looks great!  

Last thing:  I've been trying to keep the git history clean, so let's rebase the commits into 1.  

Refresher if you need it:
`git rebase -i HEAD~5`
_squish commits into the top one_
_clean commit message_
`git push -f`

bkudria commented:
All set. Thanks for the feedback!
